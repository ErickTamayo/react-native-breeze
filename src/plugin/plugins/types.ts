import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import Config from "../utils/config";

type PatterCallableParams = {
  keys: typeof Config.keys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof Config.theme;
  color: typeof Config.color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
