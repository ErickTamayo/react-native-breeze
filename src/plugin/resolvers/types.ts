export type ResolverFunctionParams = {
  input: string;
  groups: { [group: string]: string };
  theme: (key: string) => any;
};

export type ResolverFunction = (params: ResolverFunctionParams) => object;

export type Resolver = {
  pattern: RegExp;
  fn: ResolverFunction;
};
