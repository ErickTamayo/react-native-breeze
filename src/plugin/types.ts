export type Configuration = {
  theme: {
    [key: string]: any;
  };
  variants: {
    [key: string]: any;
  };
  corePlugins: {
    [key: string]: boolean;
  };
  plugins: string[];
};
