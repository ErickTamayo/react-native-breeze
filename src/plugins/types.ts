import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import { keys, colorKeys, theme, color } from "../config";

type PatterCallableParams = {
  keys: typeof keys;
  colorKeys: typeof colorKeys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof theme;
  color: typeof color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
