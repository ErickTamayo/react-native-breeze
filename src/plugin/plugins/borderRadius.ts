import { validate } from "../utils/validation";
import { PluginFunction, PluginPattern } from "./types";

export type PluginGroups = {
  position?:
    | "t"
    | "r"
    | "b"
    | "l"
    | "s"
    | "e"
    | "tl"
    | "tr"
    | "ts"
    | "te"
    | "br"
    | "bl"
    | "bs"
    | "be";
  radius?: string;
};

export const pattern: PluginPattern = ({ keys }) => {
  // prettier-ignore
  return new RegExp(`^rounded-?(?<position>t|r|b|l|s|e|tl|tr|ts|te|br|bl|bs|be)?-?(?<radius>${keys("borderRadius")})?$`);
};

export const plugin: PluginFunction<PluginGroups> = ({
  input,
  groups,
  theme,
}) => {
  const { position, radius } = groups;
  const value = theme<number>(["borderRadius", radius || "default"]);

  if (!validate(input, value, ["number"])) return {};

  switch (position) {
    case "t":
      return { borderTopRadius: value };
    case "r":
      return { borderRightRadius: value };
    case "b":
      return { borderBottomRadius: value };
    case "l":
      return { borderLeftRadius: value };
    case "s":
      return { borderStartRadius: value };
    case "e":
      return { borderEndRadius: value };
    case "tl":
      return { borderTopLeftRadius: value };
    case "tr":
      return { borderTopRightRadius: value };
    case "ts":
      return { borderTopStartRadius: value };
    case "te":
      return { borderTopEndRadius: value };
    case "bl":
      return { borderBottomLeftRadius: value };
    case "br":
      return { borderBottomRightRadius: value };
    case "bs":
      return { borderBottomStartRadius: value };
    case "be":
      return { borderBottomEndRadius: value };
    default:
      return { borderRadius: value };
  }
};
