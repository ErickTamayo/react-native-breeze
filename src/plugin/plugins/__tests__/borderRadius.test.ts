import { pattern, plugin, PluginGroups } from "../borderRadius";
import { PatternCallable, PluginFunctionReturnType } from "../types";

describe("borderRadius", () => {
  it("Should match the correct color pattern", () => {
    const theme = jest.fn().mockReturnValue({ md: 2, default: 4 });
    const callable = pattern as PatternCallable;

    expect(callable({ theme }).exec("rounded")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-md")).toBeTruthy();

    expect(callable({ theme }).exec("rounded-t")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-r")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-b")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-l")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-s")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-e")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-tl")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-tr")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-ts")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-te")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-br")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-bl")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-bs")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-be")).toBeTruthy();

    expect(callable({ theme }).exec("rounded-t-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-r-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-b-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-l-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-s-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-e-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-tl-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-tr-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-ts-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-te-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-br-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-bl-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-bs-md")).toBeTruthy();
    expect(callable({ theme }).exec("rounded-be-md")).toBeTruthy();

    expect(callable({ theme }).exec("rounded-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-t-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-r-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-b-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-l-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-tl-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-tr-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-br-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("rounded-bl-unknown")).toBeFalsy();
  });

  // prettier-ignore
  it.each<[input: string, radii: { [key: string]: number }, returned: number, expected: PluginFunctionReturnType]>([
    ["rounded", { default: 4, md: 2 }, 4, { borderRadius: 4 }],
    ["rounded-md", { default: 4, md: 2 }, 2, { borderRadius: 2 }],
    ["rounded-t", { default: 4, md: 2 }, 4, { borderTopRadius: 4 } as any], //TODO check why tf this typing is failing
  ])(
    "Should parse (%s) correctly",
    (input, radii, returned, expected) => {
      const theme = jest.fn()
      
      theme.mockReturnValueOnce(radii);
      const callable = pattern as PatternCallable;
      const groups = callable({ theme }).exec(input)!.groups! as PluginGroups;

      theme.mockReturnValueOnce(returned);
      const result = plugin({ input, groups, theme });

      expect(result).toEqual(expected);
    }
  );

  // it("Should log error if the color is not valid", () => {
  //   const theme = jest.fn().mockReturnValue({ red: { 500: 0 } });
  //   const input = "border-red-500";
  //   const callable = pattern as PatternCallable;
  //   const groups = callable({ theme }).exec(input)!.groups! as PluginGroups;
  //   const spy = jest.spyOn(console, "error").mockImplementation(() => {});

  //   const result = plugin({ input, groups, theme });

  //   expect(result).toEqual({});
  //   expect(spy).toHaveBeenCalled();

  //   spy.mockRestore();
  // });
});
