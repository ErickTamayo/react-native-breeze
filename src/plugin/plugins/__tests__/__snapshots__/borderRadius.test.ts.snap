// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`borderRadius Should parse (rounded) correctly 1`] = `
Object {
  "borderRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-b) correctly 1`] = `
Object {
  "borderBottomRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-b-sm) correctly 1`] = `
Object {
  "borderBottomRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-be) correctly 1`] = `
Object {
  "borderBottomEndRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-be-sm) correctly 1`] = `
Object {
  "borderBottomEndRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-bl) correctly 1`] = `
Object {
  "borderBottomLeftRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-bl-sm) correctly 1`] = `
Object {
  "borderBottomLeftRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-br) correctly 1`] = `
Object {
  "borderBottomRightRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-br-sm) correctly 1`] = `
Object {
  "borderBottomRightRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-bs) correctly 1`] = `
Object {
  "borderBottomStartRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-bs-sm) correctly 1`] = `
Object {
  "borderBottomStartRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-e) correctly 1`] = `
Object {
  "borderEndRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-e-sm) correctly 1`] = `
Object {
  "borderEndRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-l) correctly 1`] = `
Object {
  "borderLeftRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-l-sm) correctly 1`] = `
Object {
  "borderLeftRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-r) correctly 1`] = `
Object {
  "borderRightRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-r-sm) correctly 1`] = `
Object {
  "borderRightRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-s) correctly 1`] = `
Object {
  "borderStartRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-s-sm) correctly 1`] = `
Object {
  "borderStartRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-sm) correctly 1`] = `
Object {
  "borderRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-t) correctly 1`] = `
Object {
  "borderTopRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-t-sm) correctly 1`] = `
Object {
  "borderTopRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-te) correctly 1`] = `
Object {
  "borderTopEndRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-te-sm) correctly 1`] = `
Object {
  "borderTopEndRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-tl) correctly 1`] = `
Object {
  "borderTopLeftRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-tl-sm) correctly 1`] = `
Object {
  "borderTopLeftRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-tr) correctly 1`] = `
Object {
  "borderTopRightRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-tr-sm) correctly 1`] = `
Object {
  "borderTopRightRadius": 2,
}
`;

exports[`borderRadius Should parse (rounded-ts) correctly 1`] = `
Object {
  "borderTopStartRadius": 4,
}
`;

exports[`borderRadius Should parse (rounded-ts-sm) correctly 1`] = `
Object {
  "borderTopStartRadius": 2,
}
`;
